### ISSUE TITLE
Problem with start of road calculation with adjust_roads_and_lanes()

### ISSUE THREAD

**Issue by johschmitz** [pyoscx/scenariogeneration#122]
When connecting two roads with their end contactPoint to each other it seems that adjust_roads_and_lanes() fails. Maybe it always assumes start of successor is connected to end of predecessor or something like that?
Here is an example:
[example.zip](https://github.com/pyoscx/scenariogeneration/files/9462264/example.zip)
Either this is a bug or I overlooked/miscalculated something.

**Comment by mander76** [comment]
Can you provide a python example? :)

**Comment by johschmitz** [comment]
Sorry, zipped the wrong file, here you go:
[example.zip](https://github.com/pyoscx/scenariogeneration/files/9468067/example.zip)

**Comment by johschmitz** [comment]
I found a mistake on my side, I should have used successor instead of predecessor. However, now the heading angle still seems to be wrong. Or do you just ignore the angle supplied through the add_geometry() method?
[example_fixed.zip](https://github.com/pyoscx/scenariogeneration/files/9468195/example_fixed.zip)

**Comment by mander76** [comment]
ok, good just found that aswell.

When you add multiple roads and want to patch them together yes, then it doesn't care

The algorithm works like this: 
-checks all roads if one is fixed (so x,y, h is provided and a flat is set)
 if none are fixed, set the first road in the list of added roads
- place all other roads in relation to the first fixed road, hence creating a new x, y, h for all planviews to make a continious (including heading)

**Comment by johschmitz** [comment]
Alright that makes sense for most use-cases which use this automatic way I guess. I wanted build a non continuous road on purpose but need to make the calculations myself then.. Or do you think it would be possible to optionally(!) keep the headings and only calculate the x and y values?
Maybe if it is difficult to implement that you could at least add this information to the documentation string that it will overwrite the supplied headings. The ticket can be closed then I think.

**Comment by mander76** [comment]
you can always use the built in funcitons to get end/start point of roads/planviews. Then you just use the x and y pos from that one
`        get_end_point ()
            returns the x, y and heading at the end of the road
 `

**Comment by johschmitz** [comment]
Yes, but the idea is not to use add_fixed_geometry() but add_geometry() with the "heading" argument. Is this "heading" the same as "h_start" in the fixed geometry? Why are they not called the same?
The idea would then be, if the heading is set, the library would do some cos()/sin() calculations for me to calculate suitable x_start and y_start and ignore continuity constraints. If this should ever be implemented I think it would be good to only allow this with an **explicit** switch "disable_continuity_constraint=True" or "ignore_continuity=True" in the adjust_roads_and_lanes() method.

Probably this will never be implemented but instead maybe a warning could be emitted that the heading was changed when adjust_roads_and_lanes() was called so that people understand that it prioritizes continuity above all other constraints.

**Comment by mander76** [comment]
well, naming convensions are not the strongest suite in this package..

A warning can be fixed, don't know about the disable flag, don't really see how it would pan out in something useful

What is the use case btw? Why do you want a messed up road?

**Comment by johschmitz** [comment]
I believe you are aware of the ASAM discussions to develop tooling to scan/check/verify OpenDRIVE conformity with certain rules. This feature would allow for easier generation of OpenDRIVE XML that violates such rules and can be used as negative examples and test input for such "static analysis" tools.

**Comment by mander76** [comment]
Yeah, I sign up for that last year, but never heard anything after that? Is it on the agenda again?

Ah, true. Then I would still not make that part of the "standard" functionality of the package. Since the functionality is already there with the add_fixed_geometry().

**Comment by johschmitz** [comment]
I think this topic is moving very slow in ASAM.. What I still don't understand though is what "initial" heading will be used when calling adjust_roads_and_lanes(). Does it randomly choose one of the roads to have a fixed heading? Why don't I have to specify a road ID to be used as the fixed reference road when calling this method?

**Comment by mander76** [comment]
as I wrote on the top, if one road is fixed it will use that one, otherwise it will take the first in the first road in the all_road dict

**Comment by johschmitz** [comment]
Alright I see. First road in the dict is a bit implicit in my opinion though. It is only very recent that dicts in Python even have an order. Also if it is like that then I have to structure my consuming program in a certain way to make sure that the correct road ends up first in the dictionary. Not a problem for small examples but could be a problem for larger automatized stuff. All in all not a blocker but an extra argument like "fixed_road_id" could help some people and make the code more readable/understandable.

**Comment by mander76** [comment]
agreed, sadly have alot on my table now, but will try to fix this including the warnings soon.
